torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,39 @@
+from typing import Any
+import torch
+
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+)
+
+from torch.distributed._shard.sharded_tensor.shard import Shard
+
+
+from .metadata import (
+    STATE_DICT_TYPE,
+    MetadataIndex,
+)
+
+def _find_shard(tensor: ShardedTensor, index: MetadataIndex) -> Shard:
+    if index.offset is None:
+        raise ValueError(f"Cannot lookup {index.fqn} since its a ShardedTensor and no offset was provided")
+
+    shards = tensor.local_shards()
+    # index fast path
+    if index.index is not None:
+        if len(shards) > index.index and torch.Size(shards[index.index].metadata.shard_offsets) == index.offset:
+            return shards[index.index]
+
+    for shard in shards:
+        if torch.Size(shard.metadata.shard_offsets) == index.offset:
+            return shard
+    raise ValueError(f"Could not find shard at '{index.offset}' for FQN: '{index.fqn}'")
+
+def find_state_dict_object(state_dict: STATE_DICT_TYPE, index: MetadataIndex) -> Any:
```

Comment:
it seems like this is right now only be used for testing, is this API gonna be used by checkpointing in the next PRs?

Reply:
Yes, this API will be used by the next set of PRs that land the usage of MedataIndex. 

The whole change is 1,434 additions and 999 deletions and that doesn't include most tests, so I decided to merge it incrementally. The code is here: https://github.com/kumpera/pytorch/tree/dist_cp_new_api

I'm incrementally cleaning up, adding sufficient tests and reorganizing the prototype as I ready it for landing.